### PR TITLE
Perf testing: Cover Site Editor loading time

### LINKED ISSUE #23793: Perf testing: Cover Site Editor loading time
**Is your feature request related to a problem? Please describe.**
We've recently seen an issue (#23662) where the Site Editor took >10 seconds to load on a HyperDB-powered WordPress install that was using a distributed database. The reason was that in the PHP generating the relevant page, we were accidentally creating too many auto-drafts, which triggered HyperDB's SRTM (Send Reads To Master).

**Describe the solution you'd like**
While those are special circumstances that would easily go unnoticed on a single-database install, we realized that we could avoid similar issues by adding some basic perf test coverage to the Site Editor.

**Describe alternatives you've considered**
N/A

### PR DESCRIPTION
**Pull request by ockham**
## Description

Add some performance test coverage to the site editor (loading time), specifically to avoid perfomance regressions with regard to template resolution and `wp_template` auto-draft creation (see e.g. #23662).

Changes some of the underlying performance test framework to allow for multiple test files.

Fixes #23793.

## How has this been tested?

`npm run test-performance`

In order to run the newly added tests interactively, use `npm run test-performance -- site-editor.test.js --puppeteer-interactive`.

## Types of changes
Increase performance test coverage.

## Question

Am I doing this right? Comparing the results file to the Post Editor's, the latter's loading time is 10x the Site Editor's. This seems almost too good to be true (altough it's true that the Site Editor probably loads fewer dependencies than the Post Editor). Is it possible to run perf/e2e tests and watch what they're doing?

## Notes

Note that if you inspect this PR's 'Performance Tests' GH Action details, and expand the 'Run the performance tests' section there (and scroll to the bottom), you'll find that it actually fails upon trying to run `npm run test-performance -- packages/e2e-tests/specs/performance/post-editor.performance.test.js`. That is a file that was previously named `performance.test.js`, and is being renamed by this PR. It is being invoked by `bin/plugin/commands/performance.js`, which is also changed accordingly by this PR.

IIUC, the reason for the failure is that the GH Action is meant to compare performance for various branches, so it actually clones the `master` branch (which still has the old filenames) and runs the perf tests there (inside of a `wp-env` container that it spins up for that branch), but still uses the files from this branch to look for those test suite and results files. I don't see an easy fix, since we'd have to inject our updated files from this branch into that container. I _think_ it should be fine to merge this branch, as the GH Action should start to work properly after that.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by noahtallen — APPROVED** [review]
> Is it possible to run perf/e2e tests and watch what they're doing?

For e2e tests, yes, you just have to pass some flags. See here: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/testing-overview.md#end-to-end-testing

e.g. `npm run test-e2e:watch -- --puppeteer-interactive`

I'm not sure if that also works with the perf test, but since it uses puppeteer, I imagine it would work if you passed that flag to the right script :)

> the latter's loading time is 10x the Site Editor's

This does make sense to me since we have significantly fewer dependencies here. We also load less stuff synchronously on load with PHP. So templates are queried from WordPress _after_ the initial dom loads. (I don't think that is the case with the post editor, which probably tests it loading with a lot of blocks initially.)

> think it should be fine to merge this branch, as the GH Action should start to work properly after that.

I'm onboard with this so long as we verify that it starts working directly afterwards!

This PR looks pretty good to me 👍 I'll give a ✅ but would recommend waiting to merge for feedback from some other folks (maybe @youknowriad has some thoughts). since my knowledge of this code is limited

**Comment by noahtallen** [comment]
That said, I wonder about the test failing:
- why doesn't it cause CI to fail with a ❌ 
- is it possible to throw a more helpful error message if a file does not exist?

File: bin/plugin/commands/performance.js
Comment (youknowriad) [inline bin/plugin/commands/performance.js:225]: Should this be an argument instead of running all tests with default to post-editor? or do you think we should be running all tests everytime?

File: bin/plugin/commands/performance.js
Comment (youknowriad) [inline bin/plugin/commands/performance.js:152]: Since there's already a "performance" folder, it seems `.performance` is redundant. (same for the results)

**Review by youknowriad — CHANGES_REQUESTED** [review]
Seems like we should fix the CI before moving forward.

File: packages/e2e-tests/specs/performance/site-editor.performance.test.js
Comment (ockham) [reply packages/e2e-tests/specs/performance/site-editor.performance.test.js:75]: Yep, that's copy-pasta from the `post-editor` test. My reading was that we're basically using Jest infrastructure without really verifying behavior, but rather doing our perf measurements. (Happy to change if we can come up with something better.)

Comment (youknowriad) [reply packages/e2e-tests/specs/performance/site-editor.performance.test.js:75]: yes, tbh we can probably even remove jest entirely, the problem is that we're not using puppeteer directly but rather  relying on jest-puppeteer.

File: bin/plugin/commands/performance.js
Comment (ockham) [reply bin/plugin/commands/performance.js]: a39f46d168a855ead81c5af2f0f0cb1e3d8eab71

File: bin/plugin/commands/performance.js
Comment (ockham) [reply bin/plugin/commands/performance.js:225]: This is in the function body of `runPerformanceTests`, which is the top-level export of this file (invoked by `bin/plugin/cli.js`). If we want to make this an argument, we'd basically have to make it one that's passed by the CLI.

I'm leaning towards always running both perf suites. Their main use case is CI (the GH Action), as this is the most reliable guard against perf regressions, so it should IMO be easy to monitor that. Furthermore, the Site Editor test only covers loading time, which is currently pretty short. We can revisit if we add more coverage and testing times grow too much.

**Comment by ockham** [comment]
> Seems like we should fix the CI before moving forward.

Can you elaborate? Do you mean that it should be failing, since the perf test is throwing an error (that's being silently ignored)? I overall agree, but then I'll also need to fix the underlying issue -- which I think is only present on this branch (but will go away once merged to `master`, see PR desc). Fixing it will probably involve injecting a branch's perf test files into the repo clone that `./bin/plugin/cli.js perf` downloads (rather than using that clone's), and I'm having trouble getting that command to pass on my Linux system because of file permission issues (a known and recurring issue with `wp-env`). Long story short -- it'd be quite the blocker to sort all these things out.

As an alternative, maybe someone with macOS can run `./bin/plugin/cli.js perf` locally and confirm that it works (and post results here)?

**Comment by youknowriad** [comment]
> Can you elaborate? Do you mean that it should be failing, since the perf test is throwing an error (that's being silently ignored)?

ideally yes, at least it shouldn't fail (silently or not). In other PRs, the test seem to be passing properly. (I do see the console.log in the output)

**Comment by youknowriad** [comment]
Oh now, I understand. This is only failing because the file has been renamed in master. This makes me thing, the same will happen when we'll do `./bin/plugin/cli.js perf v8.5.0 v8.6.0` for the next release. Do you think there's a way to make it pass for old branches too somehow?

**Comment by youknowriad** [comment]
> ut still uses the files from this branch to look for those test suite and results files. I don't see an easy fix, since we'd have to inject our updated files from this branch into that container.

I think the perf command should be using master for the test and the results. At least that's how I wrote it/imagined it. That way it works regardless of the branch you're testing. The branches are only used to setup the WP+gutenberg environment.

File: bin/plugin/commands/performance.js
Comment (youknowriad) [inline bin/plugin/commands/performance.js:152]: So this seems to be the change that's causing issues in the Github action. We're changing the API of `npm run test-performance`. Can't we just leave it as is? basically run all performance tests when you do `npm run test-performance`? This is documented already in several places and I'm not sure we should be breaking this assumption?

Alternatively, we could have:

```
npm run test-performance // runs post editor tests
npm run test-performance --suite="posteditor" // runs post editor tests
npm run test-performance --suite="site-editor" // runs site editor tests
```

This will ensure we don't break the API.

Comment (ockham) [reply bin/plugin/commands/performance.js:152]: I think that's only one side of the medal though, since `master` only produces one `results.json` file for the post editor only, whereas this branch renames that file to `post-editor.results.json`, and writes the site editor results to `site-editor.results.json`.

In fact, an earlier version of this PR had `npm run test-performance` (with no added args), and that caused the GH action to fail when looking for `post-editor.results.json`, since `master` had only produced `results.json`.

I'm not sure we can retain API stability this way, _and_ add test coverage for a different test suite. We'd be extremely limited in our options, and things might end up rather hacky (something like adding site-editor specific fields to `results.json`, while retaining the post-editor ones without even renaming them -- sounds like it could become quite messy and hard to understand/maintain/extend quickly).

Overall, I'm not convinced that that's the right approach; the requirement of this kind of API stability basically means we can never significantly change perf tests, and that's just a really tough constraint, since it's really hard to anticipate whatever needs might arise in the future.

Wouldn't it make sense that a branch that changes perf tests uses _those_ tests to run on both that branch, and master?

Comment (youknowriad) [reply bin/plugin/commands/performance.js:152]: Mmm you're right, I can see that. I guess an alternative here could be to make `./bin/plugin/cli.js` take an optional "branch" where if specified, we run the perf tests from that branch instead of "master" and we could use that argument in the github action? Any thoughts on that?

Regardless, though, if we change the signature of `npm run test-performance` we should make sure to update the docs, having a `--suite` option or something like that is a bit nicer than having to pass the file.

File: packages/e2e-tests/specs/performance/site-editor.performance.test.js
Comment (ockham) [reply packages/e2e-tests/specs/performance/site-editor.performance.test.js]: Good point. What would be the best way to implement this? Add a `waitForSelector` for a thing that's only in the template? :thinking:

File: bin/plugin/commands/performance.js
Comment (ockham) [reply bin/plugin/commands/performance.js:152]: Sorry, still a few more questions :sweat_smile: 

> Mmm you're right, I can see that. I guess an alternative here could be to make `./bin/plugin/cli.js` take an optional "branch" where if specified, we run the perf tests from that branch instead of "master" and we could use that argument in the github action? Any thoughts on that?

This is actually an interface where _I_ am wondering if we need to add more options: Do we need the extra argument? Doesn't it make sense to always use the branch's tests? For most branches, those will be identical to `master`'s anyway, so it doesn't really make a difference there. But if a branch does modify them, we surely wanna use them, no?

> Regardless, though, if we change the signature of `npm run test-performance` we should make sure to update the docs, having a `--suite` option or something like that is a bit nicer than having to pass the file.

I don't mind adding that. Personally, I actually like the 'pass-through' approach (i.e. the `npm ... -- ...` syntax), mostly for its discoverability -- I didn't really look at any code but just tried it, and it did the right thing. I find that quite intuitive (and it's always the first thing I try when needing to pass a filename to an npm script). I prefer this uniform syntax over the 'nice' one, but I'm happy to go along with the 'nice' one.

---

Finally: To make the actual change, we'll have to insert the branch's test files into the wp-env container. As I said, this might be kinda painful (or even near-impossible currently) to test on my Linux box, where there's always some permissions issue when attempting to add or change files in those Docker images (and I haven't been able to fully track down that issue for a while -- wp-env works for most purposes for me, just not for stuff that requires write access, e.g. uploading media files). My alternative suggestion would be to merge without those changes, verify that the new perf test behavior works fine (and revert if it doesn't), and tackle that perf test file injection in a separate PR. Would you be okay with that? :grimacing:

Comment (youknowriad) [reply bin/plugin/commands/performance.js:152]: > Do we need the extra argument?

We need to run the same tests across branches `./bin/plugin/cli.js branchA branchB` runs the master tests in the environment built using `branchA` and `branchB` and compare both.

I was suggesting something like `./bin/plugin/cli.js branchA branchB --testsBranch branchC` to run the branchC tests in the environment built using `branchA` and `branchB` and compare both.

In the Github action BranchC will be equal to branchA though but that's fine.

> --suite config

I don't feel strongly here, the important thing is that the docs stay correct whichever choice you make.

> Finally: To make the actual change, we'll have to insert the branch's test files into the wp-env container.

I think it's just a matter of calling `git checkout branchC` at the beginning of the `perf` command (when setting up the perf environment directory)

**Review by youknowriad — COMMENTED** [review]
I think there's a mention of `npm run test-performance` in the `testing-overview.md` that needs to be updated, otherwise, this is looking good.

**Comment by ockham** [comment]
> > Is it possible to run perf/e2e tests and watch what they're doing?
> 
> For e2e tests, yes, you just have to pass some flags. See here: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/testing-overview.md#end-to-end-testing
> 
> e.g. `npm run test-e2e:watch -- --puppeteer-interactive`
> 
> I'm not sure if that also works with the perf test, but since it uses puppeteer, I imagine it would work if you passed that flag to the right script :)

Thanks! `npm run test-performance -- --puppeteer-interactive` works. I can even narrow it down to just the Site Editor tests via `npm run test-performance -- site-editor.test.js --puppeteer-interactive`.

> > the latter's loading time is 10x the Site Editor's
> 
> This does make sense to me since we have significantly fewer dependencies here. We also load less stuff synchronously on load with PHP. So templates are queried from WordPress _after_ the initial dom loads. (I don't think that is the case with the post editor, which probably tests it loading with a lot of blocks initially.)


:+1:

**Comment by youknowriad** [comment]
Capture d’écran 2020-07-15 à 8 17 58 PM

Seems like the command is failing on travis?

**Comment by ockham** [comment]
> Capture d’écran 2020-07-15 à 8 17 58 PM
> 
> Seems like the command is failing on the github action?

Yeah, I was just looking into this :confused:

I ran the tests locally in interactive mode and noticed that I need to add an `await` before `visitAdminPage` as pupeteer would otherwise simply not navigate there. I'll push the commit (but I'm not totally sure that's the reason for the error in th GH Action).

**Comment by ockham** [comment]
They're passing! :tada: 

image

I'll try removing the nasty non-numeric columns :grimacing:

**Comment by ockham** [comment]
> I'll try removing the nasty non-numeric columns

This is better: 

image

**Comment by ockham** [comment]
Should be ready for final approval :tada:

File: bin/plugin/commands/performance.js
Comment (youknowriad) [inline bin/plugin/commands/performance.js:178]: Does this has an impact on the `WPFormattedPerformanceResults` type? things should be made optional there right?

**Review by youknowriad — APPROVED** [review]

File: bin/plugin/commands/performance.js
Comment (ockham) [reply bin/plugin/commands/performance.js:178]: I thought I'm doing that here? https://github.com/WordPress/gutenberg/pull/23842/files#diff-f7011184c69b1a3bacc4458a3cc16446R51-R58 :sweat_smile: :thinking:

Comment (youknowriad) [reply bin/plugin/commands/performance.js:181]: missed this somehow 😬 😄

File: bin/plugin/commands/performance.js
Comment (dmsnell) [reply bin/plugin/commands/performance.js:154]: late to the party, but doesn't `--testsBranch` just push the problem to another location? if we change something in a PR that needs updated performance tests and then we run the tests from `trunk` then the PR will fail.

I've now spent a lot of time in this file and only today realized what `--testsBranch` is doing; I thought it did nothing.

it seems like we're trying to build consistent tests so that we can correlate runtimes over time, but I don't see a way that's possible if we also want to be able to change what we measure.

if we run the tests from each commit then we know the tests won't fail because of intentional changes, but occasionally the performance delta will look like a PR introduced a performance regression (because the underlying system changed). that's fine; these tests don't block a PR and we can investigate more granularly on the results to see if the PR slowed down particular tests we care about.

if we keep it as is I guess there's some level of consistency we gain but only by overlooking the changes we intended to make and with the occasional false-negative test crash that rejects a PR that's in reality good to go.

---

are either of you aware of a net benefit that `--testsBranch` has brought us and if we would have been any better or worse off simply running the tests in each branch/environment?

if we can get rid of this I have an idea to simplify the tests which also makes more convenient some changes that should cut out around six minutes of runtime:

 - we only ever compare two commits (we don't compare an arbitrary list of commits, so no runs of A vs. B vs. C vs. D)
 - when running on PRs for the check runs we always compare what we would merge against trunk. we can infer the appropriate commits entirely from the ENV that GitHub Actions provides.

it's still fine to have `--testsBranch` but like I said, I find it confusing, unintuitive, and I'm wondering if it actually creates more problems than it solves (or more like, it creates problems while _not_ solving the ones it was supposed to; it turns those existing problems into _different_ problems; it turns something that looked like a slowdown was introduced when it wasn't into something that rejects a PR that is good)

complicated stuff!

Comment (youknowriad) [reply bin/plugin/commands/performance.js:154]: > late to the party, but doesn't --testsBranch just push the problem to another location? if we change something in a PR that needs updated performance tests and then we run the tests from trunk then the PR will fail.

This is the exact reason why `--testsBranch` exist, to ensure that we run the updated performance tests on both branches. 

For instance, if we change code that break performance tests, we need to update in the same PR the tests to run both with the new updated code but we also need to ensure the tests continue to run properly if applied to trunk (previous code).

But also, say we change the way we compute some metric "typing" or something like that, if we decide to run different tests on different branches (say the branches tests), the results will be completely different between the branches so comparing them wouldn't make sense anymore.

Maybe I'm missing something in you reasoning, but I don't understand how `--testsBranch` is problematic.

Comment (dmsnell) [reply bin/plugin/commands/performance.js:154]: > we also need to ensure the tests continue to run properly if applied to trunk (previous code).

this was my point; is it a sound argument that we can do this? because if we change something fundamental in a branch and have to adjust the performance tests to account for it, won't `trunk` fail when running those new fixed performant tests?

we know that the test suite associated with a given branch will run; we can't say that the test suite on another branch will always run on any given different branch.

> But also, say we change the way we compute some metric "typing" or something like that, if we decide to run different tests on different branches (say the branches tests), the results will be completely different between the branches so comparing them wouldn't make sense anymore.

Yeah and I think this is just another wording of my main question. I am also having trouble understanding how `--testsBranch` solves this, because even if for a given PR we compare one set of "incompatible" tests across both branches, once we merge that PR into `trunk` I think the very next set of performance tests will now be non-comparable to all previous tests.

---

Would it be that much more trouble if, during a PR test run that shows a slowdown or speedup, that we have to subjectively investigate further the reason for the discrepancy?

If `--testsBranch` is actually working the way I think it's described as working and resolves this problem then I guess my questions are moot, but I'm still failing to understand how it does anything but shift the problem to a slightly different location (either shifting it to `trunk` after PR merge, shifting it to the target branch in PR when it was previously at the PR `HEAD`).

The only way I am currently able to understand how it helps is if we were to hypothetically create a third branch with special tests for when we make a breaking change, in order to ensure that the same tests run on both (or more) branches. That would avoid crashing the tests (which is also avoided by not having `--testsBranch`), but it leaves the same problem in `trunk` for the next PRs; the results aren't comparable to the previous commits.

---

By the way I did see that we do run a three-way perf comparison on releases so please ignore my comment about limiting to two branches. Yesterday I shuffled some code and it's fine with N branches.

Comment (youknowriad) [reply bin/plugin/commands/performance.js:154]: The rare situations where we had to make something work in different branches differently came down so far, to selector changes to open the inserter but that doesn't have an impact on any metric.

The metric is simple: - type a character, click on a block, load the page... These are the things that need to be computed similarly across branches and these things don't differ depending on the branch the test is used for. So for me personally, I think the testsBranch argument is doing its work properly so far.

Comment (dmsnell) [reply bin/plugin/commands/performance.js:154]: I will take your word for it; I don't understand what that work properly is, other than there was a PR where filenames changed and something crashed because of the lack of a tests branch. Thanks for helping clarity this! It's not a big problem that I'm seeing; was mostly just trying to remove things I don't think we need.

Comment (dmsnell) [reply bin/plugin/commands/performance.js:154]: > For instance, if we change code that break performance tests, we need to update in the same PR the tests to run both with the new updated code but we also need to ensure the tests continue to run properly if applied to trunk (previous code).

@youknowriad I'm coming back to this and trying to better understand what you wrote here; I'm wondering if you are saying that you suspect that test changes will work in a PR but then fail when run on trunk.

One thing I discovered while working on our test suites is that we don't run tests against the PR's branch, but the "PR's merge branch." The merge branch is a branch which tracks the PR but every time we make a commit it creates a hypothetical merge of that branch tip into the target branch (usually `trunk`).

What this means is that if in a PR we make changes to a test suite, then when those changes run in CI and we're relying on `$GITHUB_SHA` (as we do), then the tests already incorporate the issues of merging back into `trunk`. If the new test files would break in `trunk` they would also break on the `$GITHUB_SHA`.

I know that at that point in that PR we're comparing two different test suites, but by providing `--tests-branch` we're changing the equation from "the results in the PR comparing `trunk` to the changes are invalid" into "the results in `trunk` comparing one commit to the previous commit are invalid."

Since we only report performance changes in mainline `trunk` I'm still struggling to see how `--tests-branch` helps us achieve the goal we say it does.

cc: @ockham 

> But also, say we change the way we compute some metric "typing" or something like that, if we decide to run different tests on different branches (say the branches tests), the results will be completely different between the branches so comparing them wouldn't make sense anymore.

This is another thing I think I've been uncovering as a hidden assumption, that our results are consistent within test runs. Even with running each branch 3x I'm seeing routine variation between the branches on the perf tests, frequently hitting more than 2x difference when no difference exists. I'm going to start trying to gather some statistics on the values from individual PR runs so we can discuss the facts of the measurements we're making, but I know already for a fact that the way we talk about these measurements in the PRs doesn't match the reality I've seen.

For example, I make a change to `README.md` and in the performance tests (after 3x runs and taking the median) it shows that my branch loads the editor in half the time as it does in `trunk` and yet `trunk` is 2x faster when typing. The test code is the same, the code under test is the same, but our test runtime variance is so much wider than the time it takes to run the code we're testing that our results are probably _mostly_ random variation.

Comment (youknowriad) [reply bin/plugin/commands/performance.js:154]: > For example, I make a change to README.md and in the performance tests (after 3x runs and taking the median) it shows that my branch loads the editor in half the time as it does in trunk and yet trunk is 2x faster when typing. $

If this is happening for any of the metrics we track in https://codehealth.vercel.app then it's a problem yes. But in my experience I only saw it in site editor metrics. These are more recent and we didn't work on their stability yet so I don't think we can trust them just yet. It's just a vague indication for now.  But for the post editor metrics tracked on the site, I've not seen any major stability issues like that. I do see small variances (you can look at the graphs) but any important change has correlated well with an impactful commit.

Comment (dmsnell) [reply bin/plugin/commands/performance.js:154]: Just for some context, @youknowriad, I'm trying to start gathering data on this in #45747 and in at least the first batch of 30 test runs I found that 22 of them showed results at least as different on the `type` event (in the post editor) as [the reported change from 14.4 to 14.5](https://make.wordpress.org/core/2022/11/09/whats-new-in-gutenberg-14-5-9-november/#performance-benchmark) was - 10 of which were more than that many ms faster, and 12 of which were more than that many ms slower.

How are you measuring variance and correlation? I have to be missing something because it looks like the data is explained primarily by random variation. That is, if there is an impact, I'd like to know how you were able to pull it out of the noise caused by random variation which seems to have a bigger impact than any reported change.

In the meantime I'll spin up a PR that specifically tests that release and see if I can reproduce the measurements you found.

**Pull request merged**
